segmentstore/server/src/main/java/io/pravega/segmentstore/server/reading/StreamSegmentReadIndex.java

```diff
@@ -839,7 +873,39 @@ private CompletableReadResultEntry createRedirectedRead(long streamSegmentOffset
             // yield the right result. However, in order to recover from this without the caller's intervention, we pass
             // a pointer to getSingleReadResultEntry to the RedirectedReadResultEntry in case it fails with such an exception;
             // that class has logic in it to invoke it if needed and get the right entry.
-            result = new RedirectedReadResultEntry(result, entry.getStreamSegmentOffset(), this::getSingleReadResultEntry, this.executor);
+            result = new RedirectedReadResultEntry(result, entry.getStreamSegmentOffset(), this::getOrRegisterRedirectedRead, redirectedIndex.metadata.getId());
+        }
+
+        return result;
+    }
+
+    private CompletableReadResultEntry getOrRegisterRedirectedRead(long resultStartOffset, int maxLength, long sourceSegmentId) {
+        CompletableReadResultEntry result = getSingleReadResultEntry(resultStartOffset, maxLength);
+        if (result instanceof RedirectedReadResultEntry) {
+            // The merger isn't completed yet. Register the read so that it is completed when the merger is done.
+            PendingMerge pendingMerge;
+            synchronized (this.lock) {
+                pendingMerge = this.pendingMergers.getOrDefault(sourceSegmentId, null);
+            }
+
+            // Transform the read result entry into a Future Read, instead of adding it to futureReads, associate it
+            // with this pendingMerge, so that a completeMerge() would finalize it.
+            FutureReadResultEntry futureResult = new FutureReadResultEntry(result.getStreamSegmentOffset(), result.getRequestedReadLength());
+            if (pendingMerge != null && pendingMerge.register(futureResult)) {
+                // We were able to register the result.
+                result = futureResult;
+                log.debug("{}: Registered Pending Merge Future Read {}.", this.traceObjectId, result);
+            } else {
+                // The merge has been unregistered. Our only hope now is that the index has settled and re-invoking this
```

Comment:
This comment mentions the index settling. Is it possible that the index hasn't settled yet, in which case, I'm wondering what would happen to the read?

Reply:
This means that between line(s) 883-887 and this one, `completeMerge` has been invoked and completely executed. That means that it has sealed and unregistered the PendingMerge, so there's nothing to latch on to at this point. We can only re-issue the same call we did in 883 and get a reference to an entry in the newly merged index.

Reply:
OK. The comment gave the impression that it might not have happened, I think it is just the phrasing that confused me.